cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/nosql/NoSqlStructuredTable.java

```diff
@@ -126,6 +129,29 @@ public CloseableIterator<StructuredRow> scan(Range keyRange, int limit) throws I
                              limit);
   }
 
+  @Override
+  public CloseableIterator<StructuredRow> scan(Range keyRange, int limit, Field<?> orderByField, SortOrder sortOrder)
+    throws InvalidFieldException, IOException {
+    LOG.trace("Table {}: Scan range {} with limit {}, sort field {} and sort order {}", schema.getTableId(),
+              keyRange, limit, orderByField, sortOrder);
+    SortedIterator sortedIterator = new SortedIterator(Collections.singleton
```

Comment:
there is no way to sort on the fly while you iterate through a collection, so trying to do it in an iterator wrapper just adds unnecessary complexity. It would be better to just iterate through the scanner here, adding everything to a sorted collection, then return the iterator for that collection.

Reply:
@albertshau I had the same question, but seems like it was not done on the fly? Sorting was done in the `SortedIterator` constructor and verified in the tests. So it should be working?

Reply:
yes, it works. But the comment is about not needing the iterator

Reply:
@albertshau I think in that case we don't need this method - we can reuse https://github.com/cdapio/cdap/blob/develop/cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/nosql/NoSqlStructuredTable.java#L123 for the scan and add a method in `AppMetadataStore` https://github.com/cdapio/cdap/blob/develop/cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java for the post sorting?

Reply:
Yes but to comply with the SPI I guess we need to implement it?

Reply:
So having the sql implementation is it actually beneficial ? We scan all the rows even in that case right?

Reply:
@sanjanasandeep So I thought that we want to take advantage of the sql DB index sorting benefits when retrieving chronological versions.

Reply:
yes you are right. It is beneficial : https://www.sqlshack.com/sql-server-order-by-performance-tips/, so then let us  have the nosql implementation - just to comply with SPI

Reply:
@de-lan Should we change the return type to a `java.util.List` instead of `CloseableIterator`

Reply:
we want the CloseableIterator since the sql implementation can stream the data through instead of loading everything into memory.

Reply:
yes @de-lan and I discussed offline - we will retain `CloseableIterator`, however the nosql implementation then would have to convert the sorted collection to an iterator (which is not very efficient) since we would use a collection in `AppMetadataStore`

Reply:
@albertshau I feel like we can use the original `SortedIterator` instead of iterating through the scanner here to sort and then converting it back to an `CloseableIterator` (this is also tricky - because `ScannerIterator` requires a `Scanner` and this scanner is created from a key-range here https://github.com/cdapio/cdap/blob/develop/cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/nosql/NoSqlStructuredTable.java#L419)

Reply:
You can just iterate through and add it to a PriorityQueue like you did in the SortedIterator, it's basically just moving code.

We shouldn't create a SortedIterator because it gives the wrong impression. It pretends to be an iterator where you get one item at a time, but in reality it puts everything into memory.